analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs

```diff
@@ -42,20 +43,46 @@ public RoslynSymbolicExecution(ControlFlowGraph cfg, SymbolicExecutionCheck[] ch
 
         public void Execute()
         {
+            // ToDo: Forbid running twice
             foreach (var block in cfg.Blocks)   // ToDo: This is a temporary simplification until we support proper branching
             {
-                var state = new ProgramState(); // ToDo: This is a temporary oversimplification to scaffold the enginge
-                foreach (var operation in block.OperationsAndBranchValue.ToExecutionOrder()) // ToDo: This is a temporary oversimplification to scaffold the engine
+                queue.Enqueue(new ExplodedNode(block, ProgramState.Empty));
+                while (queue.Any())
                 {
-                    state = InvokeChecks(state, (check, ps) => check.PreProcess(ps, operation));
-                    // ToDo: Check state == null
+                    var current = queue.Dequeue();
+                    var successors = current.Operation == null ? ProcessBranching(current) : ProcessOperation(current);
```

Comment:
[edu] so for branching, the Operation is null? will it be based on block type?

Reply:
Yes, it is the same principle as in the old SE. Take a look at the `ExplodedNode.index` lifecycle. It goes `+1` in each step to process all instructions. Until it exceeds `operations.Length`, resulting in `null` being returned as Operation. Meaning that block branching should be processed.